### ISSUE TITLE
SEO Meta Titles and Descriptions update

### ISSUE THREAD

**Issue by kempsterrrr** [Developer-DAO/academy-turbo#214]
Looks like there was some awesome progress made on the SEO front from the last version though there are still some cases where the title and descriptions for the pages can be improved.

I'd like to dust of my coding chops and get a PR into academy to address these, summarised below.

- [x] Home page
    - [x] Update title `Developer DAO Academy | Home Page` > `Learn web3 with Friends | Developer DAO Academy`
    - [x] Make title a single H1 (currently it's 4)
    - [x] Add meta preview image
    - [x] Improve description - current one to short on SEO standards
- [x] Fundamentals Page
    - [x] Update title `Developer DAO Academy | Fundamentals Page` > Needs research, will update 
    - [x] improve Description - current one too short for SEO standards
    - [x] Add meta image
- [x] Fundamentals Lesson pages
    - [x] Titles not showing, update to `Lesson title | Developer DAO Academy`
    - [x] Add descriptions from lesson content
    - [ ] Review descriptions for each lesson (some are very long)
    - [x] add meta image
- [x] Tracks Page
    - [x] Update title `Developer DAO Academy | Tracks Page` > Needs research , will update
    - [x] Improve meta description - too short + add in keywords
    - [ ] ~~Review descriptions for each lesson (some are very long) (this may be out of date after discovering the `allLessonsData` file as source of truth)~~
    - [x] Add meta image
- [x] Track Page
    - [x] Update titles (showing placeholder names atm i.e. NFT Track Page
    - [x] add meta image
- [x] Lesson Page
    - [x] Update title from `Developer DAO Academy | Intro To Ethereum Track Page` > ` Intro To Ethereum | Developer DAO Academy`
    - [x] add meta image
    - [ ] ~~Add lesson descriptions as not showing~~
- [x] add sitemap.xml file ([link](https://www.npmjs.com/package/next-sitemap#getting-started))

As I'm going this I will take a close look at the SEO setup to see if there is anything that can be improved to make it more effective and easier for everyone. Will update and document any changes in the PR. This will likely include running an SEO crawl and fixing errors that come from this as well as maybe updating how the core SEO component a little.

I tried to start this work before sleeping this evening but couldn't get the local environment to work so need a little help.

**Comment by Markkos89** [comment]
hey @kempsterrrr you may want to check this notion doc to compliment remaining info https://developerdao.notion.site/S-E-O-Friendly-Titles-Meta-Descriptions-ac7f635c3c8f43f1ae02d8c3281e4cf2
just sharing the link

**Comment by kempsterrrr** [comment]
Thanks @Markkos89 ! I'm still unable to build the project locally so opened up a thread on this [in Discord here](https://discord.com/channels/883478451850473483/1135197823898501150/1212434447148912720) vs continuing in the DMs. Once I've got that sorted I'll work on this in my spare time and hope to get a PR with fixes in this week 🕺

**Comment by kempsterrrr** [comment]
Opened a PR to work through this, also copying in info from SEO audit below that helps to info some of the proposed changes.

<img width="419" alt="image" src="https://github.com/Developer-DAO/academy-turbo/assets/9025997/8ac43776-4140-4e2e-963a-fbdd31df6d57">

**Comment by kempsterrrr** [comment]
Regarding preview meta images, we have two options:

1) Design and add each on manually
2) Create a template that creates them dynamically for each page using [vercel's og-image-generation](https://vercel.com/docs/functions/og-image-generation)

Option two has potential to incur a costs though on our pro plan image generation limits are 5000/month and we're process only 77 in Jan so think we've got a long-time before we'd need to worry about this (see images below)

I propose we opt option two as this will save more time for everyone:  less design, less uploading and linking background images in code for lessons etc.

![image](https://github.com/Developer-DAO/academy-turbo/assets/9025997/ed984910-ac6e-427e-92c0-d540fcf4d4b2)
![image](https://github.com/Developer-DAO/academy-turbo/assets/9025997/8e4b083c-f282-4503-91e6-d420cb3217bb)

In practice, what we need to make this work is a template used for meta images. I propose emulate the existing designs with the image split evenly down the middle with black background and white test on the left and image on the right.

Here an example for the home page meta image

![Group 296](https://github.com/Developer-DAO/academy-turbo/assets/9025997/7d2b7ab8-458e-48ab-8368-d9a72f910f4f)

What it used to look like on social feeds

![image](https://github.com/Developer-DAO/academy-turbo/assets/9025997/ed9cddbf-5008-4a20-8e62-f1564eb53090)


What it would look like now
![image](https://github.com/Developer-DAO/academy-turbo/assets/9025997/1b4794f1-5989-4cc8-a731-79694adc0116)

These images aren't perfect as I've just thrown them tother quick but illustrates where I am going with this

**Comment by kempsterrrr** [comment]
I also need some input into how we want to populate the SEO info from the lessons. In the past, I've typically done this using values from the font matter at the top of the file - the info between the two `---`. Looks like we're currently doing this by adding props to the `<LessonLayout />` component.

Don't think it really matters which one we use. Given the lesson layout component is already implemented happy to use that and then we can just remove the font matter at the top of reach file if they're not being used.

What do you think is best @Markkos89 ?


```
---
title: Getting Started with the command line interface
description: Learn the basics of using a console/terminal/shell for Ubuntu, Mac and Windows
  WSL.
icons: []
authors: ["piablo", "georgemac510"]
---

import LessonLayout from "../../components/LessonLayout";

<LessonLayout
   lessonTitle="Command Line Interface (CLI) Basics: A Beginners Guide"
   author="piablo"
   authorImage="/authors/piablo.png"
   authorPosition="D_D Academy & Mentorship"
   authorTwitter="Skruffster"
   createdDate="August 15, 2022"
>

```

**Comment by Markkos89** [comment]
In my humble opinion I would suggest to use Next.js features the **most** we can, so in this case we could pass the title and description by props to the `LessonLayoutComponent`, and then via prop drilling we could pass those props to the `PageSEOLayout` which implements the `next-seo` package.

**Comment by Markkos89** [comment]
> Regarding preview meta images, we have two options:
> 
> 1. Design and add each on manually
> 2. Create a template that creates them dynamically for each page using [vercel's og-image-generation](https://vercel.com/docs/functions/og-image-generation)
> 
> Option two has potential to incur costs, though on our pro plan image generation limits are 5000/month and we only processed 77 in Jan so think we've got a long-time before we'd need to worry about this (see images below)
> 
> ![image](https://private-user-images.githubusercontent.com/9025997/308738930-ed984910-ac6e-427e-92c0-d540fcf4d4b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyNTAwMzAsIm5iZiI6MTcwOTI0OTczMCwicGF0aCI6Ii85MDI1OTk3LzMwODczODkzMC1lZDk4NDkxMC1hYzZlLTQyN2UtOTJjMC1kNTQwZmNmNGQ0YjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjlUMjMzNTMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWMwMWQwMDViYTM4Zjc1NTVlNzgwOWJkN2Q1ZjQ2NTUyMjVmNzRmYzdkYjhkMjBkYWI2MDUwOWJmMjYyODdiOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nbWz7h-5K14kG0VmSeQVftPHIWPNCR9_1NuYP6fopkA) ![image](https://private-user-images.githubusercontent.com/9025997/308738952-8e4b083c-f282-4503-91e6-d420cb3217bb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyNTAwMzAsIm5iZiI6MTcwOTI0OTczMCwicGF0aCI6Ii85MDI1OTk3LzMwODczODk1Mi04ZTRiMDgzYy1mMjgyLTQ1MDMtOTFlNi1kNDIwY2IzMjE3YmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjlUMjMzNTMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTk1ODYzYWJkM2NlMTAyZjU5N2FlNGQzNjIwNzc0OGMzMmVlNGRiODdmMGUyMDA4NDVlOTM0ZGMyNmY4ODY3NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.mT4i0SbMG5y6-pB6ar7ymrL8sX9j-0gT2JH2kMjSTUA)
> 
> In practice, what we need to make this work is a template used for meta images. I propose emulate the existing designs with the image split evenly down the middle with black background and white test on the left and image on the right.
> 
> Here an example for the home page meta image
> 
> ![Group 296](https://private-user-images.githubusercontent.com/9025997/308740756-7d2b7ab8-458e-48ab-8368-d9a72f910f4f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyNTAwMzAsIm5iZiI6MTcwOTI0OTczMCwicGF0aCI6Ii85MDI1OTk3LzMwODc0MDc1Ni03ZDJiN2FiOC00NThlLTQ4YWItODM2OC1kOWE3MmY5MTBmNGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjlUMjMzNTMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzVlOGQ2YWEwMzI0M2VjYmY1MTdkOWI2NzMzYjM5ZDQ4ZDM5NjQ5NmUxZDk5MDdjMmI0YTYxMzBjNjJkNWZiOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.342QaQe-FnGKX8UV8d-5XNTEJh6ZMHiSAfQ4y2BiqoE)
> 
> What it used to look like on social feeds
> 
> ![image](https://private-user-images.githubusercontent.com/9025997/308741583-ed9cddbf-5008-4a20-8e62-f1564eb53090.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyNTAwMzAsIm5iZiI6MTcwOTI0OTczMCwicGF0aCI6Ii85MDI1OTk3LzMwODc0MTU4My1lZDljZGRiZi01MDA4LTRhMjAtOGU2Mi1mMTU2NGViNTMwOTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjlUMjMzNTMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjJlM2I0OWFkZTViNTdkNmFjNmUxMzcwMTRjZWI5NTI4N2Q0NjZkZWRlNWNhMzExZTE5ZGJhYzI2MDA2MmJlYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.CYZk5CBqldSP4zTbV_ggYic0PRFB3Tzay5rxptKBQsE)
> 
> What it would look like now ![image](https://private-user-images.githubusercontent.com/9025997/308741506-1b4794f1-5989-4cc8-a731-79694adc0116.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyNTAwMzAsIm5iZiI6MTcwOTI0OTczMCwicGF0aCI6Ii85MDI1OTk3LzMwODc0MTUwNi0xYjQ3OTRmMS01OTg5LTRjYzgtYTczMS03OTY5NGFkYzAxMTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjlUMjMzNTMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzcxOWM0OGRjZTBmZWFjOTA5YjhhYWYxOWNmMzg4MmVjNGQ3ODNkNGRhNjhjODMxMjBmNzRhMDQwYTgwMDg4NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.93gyN8mRiRQWQzzGUvx2aZD_0bkquYlDtus1fYbbg1U)
> 
> These images aren't perfect as I've just thrown them tother quick but illustrates where I am going with this. @Markkos89 what do you think about adding the library og-image-generation library? Don't want to go adding libs to the codebase without checking with the 👑

Sounds great kemps! Thanks for asking you can move forward with it? Do you need any kind of help? Let me know! 
I'm about to research about the og-image-generation lib just in case. Also will take a good look to the sitemap generation

**Comment by kempsterrrr** [comment]
> Sounds great kemps! Thanks for asking you can move forward with it? Do you need any kind of help? Let me know!
> I'm about to research about the og-image-generation lib just in case. Also will take a good look to the sitemap generation

I'll find some time to have an initial pass at the og stuff and report back 🫡 - regarding sitemaps, I did some research yesterday and linked [this library](https://www.npmjs.com/package/next-sitemap#getting-started) next to the todo but I don't know if that is the best/easiest way.

Any thing you discover on the og image lib or on sitemaps lmk! happy to share the work on the pr ofc!

**Comment by kempsterrrr** [comment]
Ok I started to do the task of updating the `<LessonLayout /> component to add in the meta descriptions and then I realised there is a duplication of data across the app happening.

Some of the lesson `.mdx` files have a description in them which I can use. Some do not. I then realised there is also this file called `[allLessonData.ts](https://github.com/Developer-DAO/academy-turbo/blob/feat/seo-updates/packages/database/data/allLessonsData.ts#L263)` which has another implementation of titles and description. 

It looks like the data inside `allLessonData` is the master file for titles and description but it is not currently being used to populate the meta titles or descriptions for each individual lesson page. I need to look into the architecture more to figure if this is possible or how it could be done, but I would suggest we try to avoid have the same data in two places.

The only commit on that file is has the name `[feat: more code snippets used for seeding db](https://github.com/Developer-DAO/academy-turbo/commit/f5a1fa973233a64b60a11c62fbb8c0d2463b8eed)` so I'm assuming this means it's a seed script for populating a test database.

@Markkos89 can you help me understand this? What article data is stored in the database vs not?

In my experience building markdown powered content sites. The metadata for each piece of content is stored in the font-matter at the top of the md/mdx file and then this is data is used to populate the titles, descriptions etc. for each piece of content. This keeps the all info the the content in one place vs some in mdx files in the repo and some living in a database.

Might be good to have a call to understand the direction you've chosen here better before I move forward and break things :) 

My cal link - https://cal.com/kempsterrrr

**Comment by Markkos89** [comment]
What you see on `allLessonData` (and `allTracksDaa`, etc) are just the first **export** from the database to have them as files. Was part of changing from static arrays to start using the database and now we store the latest updated info in the database.

**Comment by kempsterrrr** [comment]
Ok, so the source of truth for article meta data (title, description, authors etc.) is now stored in the DB? If yes, how is the info loaded into the database

**Comment by Markkos89** [comment]
> Ok, so the source of truth for article meta data (title, description, authors etc.) is now stored in the DB? If yes, how is the info loaded into the database cc @Markkos89 and kept up to date/edited

For now is being populated hand-made via Prisma DB seed scripts: https://developerdao.notion.site/DB-Seed-Scripts-aa34eaf9f08440399596b577075fa779

> ..and if the source of truth for the data is stored there it makes most sense to populate the meta data for <LessonLayout /> component from the DB too, rather than hard coding it into each lesson .mdx file

Yes, totally agree. That way we also are going to be able to apply some optimization techniques using Next.js features

**Comment by kempsterrrr** [comment]
ok yea, so I can't figure out a way to pipe the meta data from the db into the `<LessonLayout />` component for each lesson with needing to duplicate at least one piece of data in the local .mdx file.

I believe I can bring in the DB data from `allLessonData` into the `<LessonLayout />` component via the `<AppContext />` but if I do that I'll need some way to filter the data to only provide the data for each individual lesson and as far as I can tell the only way to do that would be to hard code a unique value into the mdx file that matches a unique value in the db (`allLessonData`) which means we have the duplicated data problem as we're trying to combine remote and local data (used to having all data for a page coming from one place - local file / headless cms)

I'm probably missing something here re the architecture and how mdx files are being rendered so love to chat on this before moving ahead with any more updates.

**Comment by kempsterrrr** [comment]
Sharing an update here. Finalise the pr here #215 

From conversations with @Markkos89 we decided to ship a reduced version of this PR. What it gives us is meta titles, descriptions and images for each page of the site with the following compromises.

1. We're using the same meta image on every page when we ideally want unique meta images for most pages
2. We're using a default meta description for the lessons rather than unique ones as everything is hardcoded and wasn't sure it was worth the effort to do this update if the data will be coming from the DB soon

I would also recommend changes to some of the lesson titles and descriptions to make them more SEO friendly / more aligned with both keywords and search terms that user search on but that felt out of scope of this PR so I didn't push any of these :)

I think it's probably better to open two new issues to for each of the points of above so we can keep track of them indapendantly vs keeping this big issue open. wdyt @Markkos89 ?

**Comment by Markkos89** [comment]
> Sharing an update here. Finalise the pr here #215
> 
> From conversations with @Markkos89 we decided to ship a reduced version of this PR. What it gives us is meta titles, descriptions and images for each page of the site with the following compromises.
> 
> 1. We're using the same meta image on every page when we ideally want unique meta images for most pages
> 2. We're using a default meta description for the lessons rather than unique ones as everything is hardcoded and wasn't sure it was worth the effort to do this update if the data will be coming from the DB soon
> 
> I would also recommend changes to some of the lesson titles and descriptions to make them more SEO friendly / more aligned with both keywords and search terms that user search on but that felt out of scope of this PR so I didn't push any of these :)
> 
> I think it's probably better to open two new issues to for each of the points of above so we can keep track of them indapendantly vs keeping this big issue open. wdyt @Markkos89 ?

Amazing! Yeah I also agree to open two new issues.. you can tag them there after created and close this one, so we have the reference with a text flag like `related to #333` etc

Amazing contribution Kemps, thank you very much :raised_hands:

**Comment by kempsterrrr** [comment]
Done, feel free to close this once #215 is merged in 🚀

**Comment by Markkos89** [comment]
> Done, feel free to close this once #215 is merged in 🚀

mate this should goin to main/production, right?

**Comment by Markkos89** [comment]
Closing this issue as the update was successfully **merged to production**
